es_out: support Upstream Servers with configuration overriding (#1143)
fluent/fluent-bit-docs: pipeline/outputs/elasticsearch.md

## Configuration Parameters

> | Key | Description | default | Overridable in NODE section of [Upstream](../../administration/configuring-fluent-bit/classic-mode/upstream-servers.md) configuration |
> | :--- | :--- | :--- | :--- |
> | Host | IP address or hostname of the target Elasticsearch instance | 127.0.0.1 | Yes, default value is not applicable for NODE section of Upstream configuration, which **requires** host to be specified |
> | Port | TCP port of the target Elasticsearch instance | 9200 | Yes, default value is not applicable for NODE section of Upstream configuration, which **requires** port to be specified |
> | Path | Elasticsearch accepts new data on HTTP query path "/\_bulk". But it is also possible to serve Elasticsearch behind a reverse proxy on a subpath. This option defines such path on the fluent-bit side. It simply adds a path prefix in the indexing HTTP POST URI. | Empty string | Yes |
> | compress | Set payload compression mechanism. Option available is 'gzip' | | Yes |
> | Buffer\_Size | Specify the buffer size used to read the response from the Elasticsearch HTTP service. This option is useful for debugging purposes where is required to read full responses, note that response size grows depending of the number of records inserted. To set an _unlimited_ amount of memory set this value to **False**, otherwise the value must be according to the [Unit Size](../../administration/configuring-fluent-bit/unit-sizes.md) specification. | 4KB | Yes |
> | Pipeline | Newer versions of Elasticsearch allows to setup filters called pipelines. This option allows to define which pipeline the database should use. For performance reasons is strongly suggested to do parsing and filtering on Fluent Bit side, avoid pipelines. |  | Yes |
> | AWS\_Auth | Enable AWS Sigv4 Authentication for Amazon OpenSearch Service | Off | Yes |
> | AWS\_Region | Specify the AWS region for Amazon OpenSearch Service |  | Yes |

Reviewer: ```suggestion
The **Overridable** column indicates if a key can be overridden in the NODE section
of an
[Upstream](../../administration/configuring-fluent-bit/classic-mode/upstream-servers.md)
configuration.

| Key  | Description | Default | Overridable |
| :--- | :--- | :--- | :--- |
| Host | IP address or hostname of the target Elasticsearch instance. | `127.0.0.1` | Yes. Default value isn't applicable for NODE section of Upstream configuration, which **requires** Host to be specified. |
| Port | TCP port of the target Elasticsearch instance | `9200` | Yes. Default value isn't applicable for NODE section of Upstream configuration, which **requires** Port to be specified. |
| Path | Elasticsearch accepts new data on HTTP query path `/_bulk`. But it is also possible to serve Elasticsearch behind a reverse proxy on a subpath. This option defines such path on the fluent-bit side. It simply adds a path prefix in the indexing HTTP POST URI. | Empty string | Yes |
| compress | Set payload compression mechanism. Option available is `gzip`. | _none_ | Yes |
| Buffer\_Size | Specify the buffer size used to read the response from the Elasticsearch HTTP service. This option is useful for debugging purposes where is required to read full responses, note that response size grows depending of the number of records inserted. To set an _unlimited_ amount of memory set this value to **False**, otherwise the value must be according to the [Unit Size](../../administration/configuring-fluent-bit/unit-sizes.md) specification. | `4KB` | Yes |
| Pipeline | Newer versions of Elasticsearch allows to setup filters called pipelines. This option allows to define which pipeline the database should use. For performance reasons is strongly suggested to do parsing and filtering on Fluent Bit side, avoid pipelines. | _none_ | Yes |
| AWS\_Auth | Enable AWS Sigv4 Authentication for Amazon OpenSearch Service | `Off` | Yes |
| AWS\_Region | Specify the AWS region for Amazon OpenSearch Service | _none_ | Yes |
```

Author: Suggestion with modification - `Off` and `On` can be considered special values and not literals, so `Off` and `On` are not wrapped with backticks, `Empty string` is replaced with `_empty string_` for consistency with `_none_` - was applied in c620db5a2ca5a0cd0f5a6474e0db8a0139bca3c3, cdde1dec1e45673d9a9554429feea6df14f113c0 and 929d05ddfc37113904775d9e80e0d6f1fefcb1a9.

> | AWS\_STS\_Endpoint | Specify the custom sts endpoint to be used with STS API for Amazon OpenSearch Service |  | Yes |
> | AWS\_Role\_ARN | AWS IAM Role to assume to put records to your Amazon cluster |  | Yes |
> | AWS\_External\_ID | External ID for the AWS IAM Role specified with `aws_role_arn` |  | Yes |
> | AWS\_Service\_Name | Service name to be used in AWS Sigv4 signature. For integration with Amazon OpenSearch Serverless, set to `aoss`. See the [FAQ](opensearch.md#faq) section on Amazon OpenSearch Serverless for more information. | es | Yes |
> | AWS\_Profile | AWS profile name | default | Yes |
> | Cloud\_ID | If you are using Elastic's Elasticsearch Service you can specify the cloud\_id of the cluster running. The Cloud ID string has the format `<deployment_name>:<base64_info>`. Once decoded, the `base64_info` string has the format `<deployment_region>$<elasticsearch_hostname>$<kibana_hostname>`. |  | No |
> | Cloud\_Auth | Specify the credentials to use to connect to Elastic's Elasticsearch Service running on Elastic Cloud |  | Yes |
> | HTTP\_User | Optional username credential for Elastic X-Pack access |  | Yes |
> | HTTP\_Passwd | Password for user defined in HTTP\_User |  | Yes |
> | Index | Index name | fluent-bit | Yes |

Reviewer: ```suggestion
| AWS\_STS\_Endpoint | Specify the custom sts endpoint to be used with STS API for Amazon OpenSearch Service. | _none_ | Yes |
| AWS\_Role\_ARN | AWS IAM Role to assume to put records to your Amazon cluster. | _none_ | Yes |
| AWS\_External\_ID | External ID for the AWS IAM Role specified with `aws_role_arn`. | _none_ | Yes |
| AWS\_Service\_Name | Service name to be used in AWS Sigv4 signature. For integration with Amazon OpenSearch Serverless, set to `aoss`. See the [FAQ](opensearch.md#faq) section on Amazon OpenSearch Serverless for more information. | `es` | Yes |
| AWS\_Profile | AWS profile name. | `default` | Yes |
| Cloud\_ID | If you are using Elastic's Elasticsearch Service you can specify the `cloud_id` of the cluster running. The Cloud ID string has the format `<deployment_name>:<base64_info>`. After decoding, the `base64_info` string has the format `<deployment_region>$<elasticsearch_hostname>$<kibana_hostname>`. | _none_ | No |
| Cloud\_Auth | Specify the credentials to use to connect to Elastic's Elasticsearch Service running on Elastic Cloud. | _none_ | Yes |
| HTTP\_User | Optional username credential for Elastic X-Pack access. | _none_ | Yes |
| HTTP\_Passwd | Password for user defined in HTTP\_User. | _none_ | Yes |
| Index | Index name. | `fluent-bit` | Yes |
```

Author: Hi @cnorris-cs,

I'm sorry, but it doesn't look like other parts of Fluent Bit documentation (like https://github.com/fluent/fluent-bit-docs/blob/master/pipeline/outputs/forward.md) wrap default values with backticks, i.e. the parts of your suggestion like replacing `fluent-bit` with 

```md
`fluent-bit`
```

seem to not match existing documentation style.

Same for:

```md
_none_
```

in "Default" column for the keys without default value - the most of existing documentation keeps such cells empty.

Please note also that values in "Description" and "Default" columns were mostly copied from original version of this file, so I'm not sure if this pull request is allowed to change these values (because this pull request is about overriding only).

Thank you.

Reviewer: We're actually going to be working on updating the documentation to be along the lines of what I suggested. Empty cells don't always communicate the concept of "nothing". Values to be entered should be consistent, and marking them as code help to signify that they're items to be entered.

So, we're making these suggestions to help us get where we want to be with the docs, knowing that it'll take some time to get there. But moving in that direction sooner than later helps with all of this.

(And most of the changes to the description column were to do things like add punctuation and to incorporate grammar changes. I do know of one line that I did some work on that would need review -- the ElasticSearch line -- but it would benefit from some changes for clarity.)

> | Type | Type name | \_doc | Yes |
> | Logstash\_Format | Enable Logstash format compatibility. This option takes a boolean value: True/False, On/Off | Off | Yes |
> | Logstash\_Prefix | When Logstash\_Format is enabled, the Index name is composed using a prefix and the date, e.g: If Logstash\_Prefix is equals to 'mydata' your index will become 'mydata-YYYY.MM.DD'. The last string appended belongs to the date when the data is being generated. | logstash | Yes |
> | Logstash\_Prefix\_Key | When included: the value of the key in the record will be evaluated as key reference and overrides Logstash\_Prefix for index generation. If the key/value is not found in the record then the Logstash\_Prefix option will act as a fallback. The parameter is expected to be a [record accessor](../../administration/configuring-fluent-bit/classic-mode/record-accessor.md). |  | Yes |
> | Logstash\_Prefix\_Separator | Set a separator between logstash_prefix and date.| - | Yes |
> | Logstash\_DateFormat | Time format \(based on [strftime](http://man7.org/linux/man-pages/man3/strftime.3.html)\) to generate the second part of the Index name. | %Y.%m.%d | Yes |
> | Time\_Key | When Logstash\_Format is enabled, each record will get a new timestamp field. The Time\_Key property defines the name of that field. | @timestamp | Yes |
> | Time\_Key\_Format | When Logstash\_Format is enabled, this property defines the format of the timestamp. | %Y-%m-%dT%H:%M:%S | Yes |
> | Time\_Key\_Nanos | When Logstash\_Format is enabled, enabling this property sends nanosecond precision timestamps. | Off | Yes |
> | Include\_Tag\_Key | When enabled, it append the Tag name to the record. | Off | Yes |

Reviewer: ```suggestion
| Type | Type name | `\_doc` | Yes |
| Logstash\_Format | Enable Logstash format compatibility. This option takes a boolean value: True/False, On/Off. | `Off` | Yes |
| Logstash\_Prefix | When Logstash\_Format is enabled, the Index name is composed using a prefix and the date. For example, if Logstash\_Prefix is `mydata`, your index becomes `mydata-YYYY.MM.DD`. The last string appended belongs to the date when the data is being generated. | `logstash` | Yes |
| Logstash\_Prefix\_Key | When included, the value of the key in the record will be evaluated as key reference and overrides Logstash\_Prefix for index generation. If the key/value isn't found in the record, the Logstash\_Prefix option will act as a fallback. The parameter is expected to be a [record accessor](../../administration/configuring-fluent-bit/classic-mode/record-accessor.md). | _none_ | Yes |
| Logstash\_Prefix\_Separator | Set a separator between logstash_prefix and date. | `-` | Yes |
| Logstash\_DateFormat | Time format \(based on [strftime](http://man7.org/linux/man-pages/man3/strftime.3.html)\) to generate the second part of the Index name. | `%Y.%m.%d` | Yes |
| Time\_Key | When Logstash\_Format is enabled, each record gets a new timestamp field. The Time\_Key property defines the name of that field. | `@timestamp` | Yes |
| Time\_Key\_Format | When Logstash\_Format is enabled, this property defines the format of the timestamp. | `%Y-%m-%dT%H:%M:%S` | Yes |
| Time\_Key\_Nanos | When Logstash\_Format is enabled, enabling this property sends nanosecond precision timestamps. | `Off` | Yes |
| Include\_Tag\_Key | When enabled, it append the Tag name to the record. | `Off` | Yes |
```

Author: Suggestion with small fix (there is no need to escape underscore with backslash if underscore is wrapped with backticks) and modification (`Off` and `On` can be considered special values and not literals, so `Off` and `On` are not wrapped with backticks) was applied in c620db5a2ca5a0cd0f5a6474e0db8a0139bca3c3, cdde1dec1e45673d9a9554429feea6df14f113c0 and 929d05ddfc37113904775d9e80e0d6f1fefcb1a9.

> | Tag\_Key | When Include\_Tag\_Key is enabled, this property defines the key name for the tag. | \_flb-key | Yes |
> | Generate\_ID | When enabled, generate `_id` for outgoing records. This prevents duplicate records when retrying ES. | Off | Yes |
> | Id\_Key | If set, `_id` will be the value of the key from incoming record and `Generate_ID` option is ignored. |  | Yes |
> | Write\_Operation | The write\_operation can be any of: create (default), index, update, upsert. | create | Yes |
> | Replace\_Dots | When enabled, replace field name dots with underscore, required by Elasticsearch 2.0-2.3. | Off | Yes |
> | Trace\_Output | Print all elasticsearch API request payloads to stdout \(for diag only\) | Off | Yes |
> | Trace\_Error | If elasticsearch return an error, print the elasticsearch API request and response \(for diag only\) | Off | Yes |
> | Current\_Time\_Index | Use current time for index generation instead of message record | Off | Yes |
> | Suppress\_Type\_Name | When enabled, mapping types is removed and `Type` option is ignored. If using Elasticsearch 8.0.0 or higher - it [no longer supports mapping types](https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html), so it shall be set to On. | Off | Yes |
> | Workers | The number of [workers](../../administration/multithreading.md#outputs) to perform flush operations for this output. | `2` | No |
> | Upstream | If plugin will connect to an _Upstream_ instead of a simple host, this property defines the absolute path for the Upstream configuration file, for more details about this refer to the [Upstream Servers](../../administration/configuring-fluent-bit/classic-mode/upstream-servers.md) documentation section. | | No |

Reviewer: ```suggestion
| Tag\_Key | When Include\_Tag\_Key is enabled, this property defines the key name for the tag. | `\_flb-key` | Yes |
| Generate\_ID | When enabled, generate `_id` for outgoing records. This prevents duplicate records when retrying ES. | `Off` | Yes |
| Id\_Key | If set, `_id` is the value of the key from incoming record, and `Generate_ID` option is ignored. | _none_ | Yes |
| Write\_Operation | The write\_operation can be any of: `create`, `index`, `update`, `upsert`. | `create` | Yes |
| Replace\_Dots | When enabled, replace field name dots with underscore, required by Elasticsearch 2.0-2.3. | `Off` | Yes |
| Trace\_Output | Print all elasticsearch API request payloads to stdout. For diagnostic use only. | `Off` | Yes |
| Trace\_Error | If elasticsearch return an error, print the elasticsearch API request and response. For diagnostic use only. | `Off` | Yes |
| Current\_Time\_Index | Use current time for index generation instead of message record. | `Off` | Yes |
| Suppress\_Type\_Name | When enabled, mapping types is removed and `Type` option is ignored. Elasticsearch 8.0.0 or later [doesn't support mapping types](https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html), which requires this value to be `On`. | `Off` | Yes |
| Workers | The number of [workers](../../administration/multithreading.md#outputs) to perform flush operations for this output. | `2` | No |
| Upstream | If plugin will connect to an _Upstream_ instead of a simple host, this property defines the absolute path for the Upstream configuration file, for more details about this refer to the [Upstream Servers](../../administration/configuring-fluent-bit/classic-mode/upstream-servers.md) documentation section. | _none_ | No |
```

Author: Suggestion with small fix (there is no need to escape underscore with backslash if underscore is wrapped with backticks) and modification (`Off` and `On` can be considered special values and not literals, so there are not wrapped with backticks) was applied in c620db5a2ca5a0cd0f5a6474e0db8a0139bca3c3.

> The parameters _index_ and _type_ can be confusing if you are new to Elastic, if you have used a common relational database before, they can be compared to the _database_ and _table_ concepts. Also see [the FAQ below](elasticsearch.md#faq)

### TLS / SSL

Elasticsearch output plugin supports TLS/SSL, for more details about the properties available and general configuration, please refer to the [TLS/SSL](../../administration/transport-security.md) section.

### AWS Sigv4 Authentication and Upstream Servers

> http_proxy, no_proxy and TLS parameters used for AWS Sigv4 Authentication - for connection of plugin to AWS to generate authentication signature - are never picked from NODE section of [Upstream](../../administration/configuring-fluent-bit/classic-mode/upstream-servers.md) configuration.
> TLS parameters for connection of plugin to Elasticsearch **can** be overridden in NODE section of Upstream (even if AWS authentication is used).

Reviewer: ```suggestion
The `http_proxy`, `no_proxy`, and `TLS` parameters used for AWS Sigv4 Authentication
(for connection of plugin to AWS to generate authentication signature) are never
picked from the `NODE` section of the
[Upstream](../../administration/configuring-fluent-bit/classic-mode/upstream-servers.md)
configuration. However, `TLS` parameters for connection of the plugin to
Elasticsearch **can** be overridden in the `NODE` section of Upstream, even if AWS
authentication is used.
```

Author: Suggestion was applied as part of f28a9c6a402fcf5dc87b1aed0902e1dff09c6749.

| `Id_Key` | If set, `_id` is the value of the key from incoming record, and `Generate_ID` option is ignored. | _none_ | Yes |
| `Write_Operation` | `Write_operation` can be any of: `create`, `index`, `update`, `upsert`. | `create` | Yes |
| `Replace_Dots` | When enabled, replace field name dots with underscore. Required by Elasticsearch 2.0-2.3. | `Off` | Yes |
| `Trace_Output` | Print all ElasticSearch API request payloads to `stdout` for diagnostics. | `Off` | Yes |
| `Trace_Error` | If ElasticSearch returns an error, print the ElasticSearch API request and response for diagnostics. | `Off` | Yes |
| `Current_Time_Index` | Use current time for index generation instead of message record. | `Off` | Yes |
| `Suppress_Type_Name` | When enabled, mapping types is removed and `Type` option is ignored. Elasticsearch 8.0.0 or later [doesn't support mapping types](https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html), which requires this value to be `On`. | `Off` | Yes |
| `Workers` | The number of [workers](../../administration/multithreading.md#outputs) to perform flush operations for this output. | `2` | No |

> | `Upstream` | If plugin will connect to an _Upstream_ instead of a simple host, this property defines the absolute path for the Upstream configuration file, for more details about this refer to the [Upstream Servers](../../administration/configuring-fluent-bit/classic-mode/upstream-servers.md) documentation section. | _none_ | No |

Reviewer 2: Is this the only new addition? With the extra column it makes it hard to check

Author: Yes, mostly new column `Overridable` was added. Here is the diff without new column:
```diff
 | Key | Description | Default |
 | :--- | :--- | :--- |
-| `Host` | IP address or hostname of the target Elasticsearch instance | `127.0.0.1` |
-| `Port` | TCP port of the target Elasticsearch instance | `9200` |
+| `Host` | IP address or hostname of the target Elasticsearch instance. | `127.0.0.1` |
+| `Port` | TCP port of the target Elasticsearch instance. | `9200` |
 | `Path` | Elasticsearch accepts new data on HTTP query path `/_bulk`. You can also serve Elasticsearch behind a reverse proxy on a sub-path. Define the path by adding a path prefix in the indexing HTTP POST URI. | Empty string |
 | `compress` | Set payload compression mechanism. Option available is `gzip`. | _none_ |
 | `Buffer_Size` | Specify the buffer size used to read the response from the Elasticsearch HTTP service. Use for debugging purposes where required to read full responses. Response size grows depending of the number of records inserted. To use an unlimited amount of memory, set this value to `False`. Otherwise set the value according to the [Unit Size](../../administration/configuring-fluent-bit/unit-sizes.md). | `512KB` |
 | `Pipeline` | Define which pipeline the database should use. For performance reasons, it's strongly suggested to do parsing and filtering on Fluent Bit side, and avoid pipelines. | _none_ |
 | `AWS_Auth` | Enable AWS Sigv4 Authentication for Amazon OpenSearch Service. | `Off` |
 | `AWS_Region` | Specify the AWS region for Amazon OpenSearch Service. | _none_ |
-| `AWS_STS_Endpoint` | Specify the custom STS endpoint to be used with STS API for Amazon OpenSearch Service | _none_ |
-| `AWS_Role_ARN` | AWS IAM Role to assume to put records to your Amazon cluster | _none_ |
-| `AWS_External_ID` | External ID for the AWS IAM Role specified with `aws_role_arn` | _none_ |
+| `AWS_STS_Endpoint` | Specify the custom STS endpoint to be used with STS API for Amazon OpenSearch Service. | _none_ |
+| `AWS_Role_ARN` | AWS IAM Role to assume to put records to your Amazon cluster. | _none_ |
+| `AWS_External_ID` | External ID for the AWS IAM Role specified with `aws_role_arn`. | _none_ |
 | `AWS_Service_Name` | Service name to use in AWS Sigv4 signature. For integration with Amazon OpenSearch Serverless, set to `aoss`. See [Amazon OpenSearch Serverless](opensearch.md) for more information. | `es` |
-| `AWS_Profile` | AWS profile name | `default` |
-| `Cloud_ID` | If using Elastic's Elasticsearch Service you can specify the `cloud_id` of the cluster running. The string has the format `<deployment_name>:<base64_info>`. Once decoded, the `base64_info` string has the format `<deployment_region>$<elasticsearch_hostname>$<kibana_hostname>`. | _none_ |
-| `Cloud_Auth` | Specify the credentials to use to connect to Elastic's Elasticsearch Service running on Elastic Cloud | _none_ |
-| `HTTP_User` | Optional username credential for Elastic X-Pack access | _none_ |
-| `HTTP_Passwd` | Password for user defined in `HTTP_User` | _none_ |
-| `Index` | Index name | `fluent-bit` |
-| `Type` | Type name | `_doc` |
-| `Logstash_Format` | Enable Logstash format compatibility. This option takes a Boolean value: `True/False`, `On/Off` | `Off` |
-| `Logstash_Prefix` | When `Logstash_Format` is enabled, the Index name is composed using a prefix and the date, e.g: If `Logstash_Prefix` is equal to `mydata` your index will become `mydata-YYYY.MM.DD`. The last string appended belongs to the date when the data is being generated. | `logstash` |
-| `Logstash_Prefix_Key` | When included: the value of the key in the record will be evaluated as key reference and overrides `Logstash_Prefix` for index generation. If the key/value isn't found in the record then the `Logstash_Prefix` option will act as a fallback. The parameter is expected to be a [record accessor](../../administration/configuring-fluent-bit/classic-mode/record-accessor.md). | _none_ |
-| `Logstash_Prefix_Separator` | Set a separator between `Logstash_Prefix` and date.| `-` |
+| `AWS_Profile` | AWS profile name. | `default` |
+| `Cloud_ID` | If using Elastic's Elasticsearch Service you can specify the `cloud_id` of the cluster running. The string has the format `<deployment_name>:<base64_info>`. After decoding, the `base64_info` string has the format `<deployment_region>$<elasticsearch_hostname>$<kibana_hostname>`. | _none_ |
+| `Cloud_Auth` | Specify the credentials to use to connect to Elastic's Elasticsearch Service running on Elastic Cloud. | _none_ |
+| `HTTP_User` | Optional username credential for Elastic X-Pack access. | _none_ |
+| `HTTP_Passwd` | Password for user defined in `HTTP_User`. | _none_ |
+| `Index` | Index name. | `fluent-bit` |
+| `Type` | Type name. | `_doc` |
+| `Logstash_Format` | Enable Logstash format compatibility. This option takes a Boolean value: `True/False`, `On/Off`. | `Off` |
+| `Logstash_Prefix` | When `Logstash_Format` is enabled, the Index name is composed using a prefix and the date, For example, if `Logstash_Prefix` is equal to `mydata`, your index becomes `mydata-YYYY.MM.DD`. The last string appended belongs to the date when the data is being generated. | `logstash` |
+| `Logstash_Prefix_Key` | When included: the value of the key in the record will be evaluated as key reference and overrides `Logstash_Prefix` for index generation. If the key/value isn't found in the record, the `Logstash_Prefix` option will act as a fallback. The parameter is expected to be a [record accessor](../../administration/configuring-fluent-bit/classic-mode/record-accessor.md). | _none_ |
+| `Logstash_Prefix_Separator` | Set a separator between `Logstash_Prefix` and date. | `-` |
 | `Logstash_DateFormat` | Time format based on [strftime](http://man7.org/linux/man-pages/man3/strftime.3.html) to generate the second part of the Index name. | `%Y.%m.%d` |
-| `Time_Key` | When `Logstash_Format` is enabled, each record will get a new timestamp field. The `Time_Key` property defines the name of that field. | `@timestamp` |
+| `Time_Key` | When `Logstash_Format` is enabled, each record gets a new timestamp field. The `Time_Key` property defines the name of that field. | `@timestamp` |
 | `Time_Key_Format` | When `Logstash_Format` is enabled, this property defines the format of the timestamp. | `%Y-%m-%dT%H:%M:%S` |
 | `Time_Key_Nanos` | When `Logstash_Format` is enabled, enabling this property sends nanosecond precision timestamps. | `Off` |
 | `Include_Tag_Key` | When enabled, it append the Tag name to the record. | `Off` |
 | `Tag_Key` | When `Include_Tag_Key` is enabled, this property defines the key name for the tag. | `_flb-key` |
 | `Generate_ID` | When enabled, generate `_id` for outgoing records. This prevents duplicate records when retrying ES. | `Off` |
-| `Id_Key` | If set, `_id` will be the value of the key from incoming record and `Generate_ID` option is ignored. | _none_ |
+| `Id_Key` | If set, `_id` is the value of the key from incoming record, and `Generate_ID` option is ignored. | _none_ |
 | `Write_Operation` | `Write_operation` can be any of: `create`, `index`, `update`, `upsert`. | `create` |
 | `Replace_Dots` | When enabled, replace field name dots with underscore. Required by Elasticsearch 2.0-2.3. | `Off` |
 | `Trace_Output` | Print all ElasticSearch API request payloads to `stdout` for diagnostics. | `Off` |
 | `Trace_Error` | If ElasticSearch returns an error, print the ElasticSearch API request and response for diagnostics. | `Off` |
 | `Current_Time_Index` | Use current time for index generation instead of message record. | `Off` |
-| `Suppress_Type_Name` | When enabled, mapping types is removed and `Type` option is ignored. Elasticsearch 8.0.0 or higher [no longer supports mapping types](https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html), and is set to `On`. | `Off` |
+| `Suppress_Type_Name` | When enabled, mapping types is removed and `Type` option is ignored. Elasticsearch 8.0.0 or later [doesn't support mapping types](https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html), which requires this value to be `On`. | `Off` |
 | `Workers` | The number of [workers](../../administration/multithreading.md#outputs) to perform flush operations for this output. | `2` |
+| `Upstream` | If plugin will connect to an _Upstream_ instead of a simple host, this property defines the absolute path for the Upstream configuration file, for more details about this refer to the [Upstream Servers](../../administration/configuring-fluent-bit/classic-mode/upstream-servers.md) documentation section. | _none_ |
```

This plugin has the following configuration parameters:

The **Overridable** column indicates if a key can be overridden in the NODE section of an
[Upstream](../../administration/configuring-fluent-bit/classic-mode/upstream-servers.md)
configuration.

| Key | Description | Default | Overridable |
| :--- | :--- | :--- | :--- |

> | `Host` | IP address or hostname of the target Elasticsearch instance. | `127.0.0.1` | Yes. Default value isn't applicable for NODE section of Upstream configuration, which **requires** Host to be specified. |
> | `Port` | TCP port of the target Elasticsearch instance. | `9200` | Yes. Default value isn't applicable for NODE section of Upstream configuration, which **requires** Port to be specified. |

Reviewer 3: ```suggestion
| `Host` | IP address or hostname of the target Elasticsearch instance. | `127.0.0.1` | Yes. Default value isn't applicable for NODE section of Upstream configuration, which requires `Host` to be specified. |
| `Port` | TCP port of the target Elasticsearch instance. | `9200` | Yes. Default value isn't applicable for NODE section of Upstream configuration, which requires `Port` to be specified. |
```

We don't use bold for emphasis, per our style guide. Also ensuring keys are code fenced.

Author: Applied these changes

| `Suppress_Type_Name` | When enabled, mapping types is removed and `Type` option is ignored. Elasticsearch 8.0.0 or later [doesn't support mapping types](https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html), which requires this value to be `On`. | `Off` | Yes |
| `Workers` | The number of [workers](../../administration/multithreading.md#outputs) to perform flush operations for this output. | `2` | No |
| `Upstream` | If plugin will connect to an _Upstream_ instead of a simple host, this property defines the absolute path for the Upstream configuration file, for more details about this refer to the [Upstream Servers](../../administration/configuring-fluent-bit/classic-mode/upstream-servers.md) documentation section. | _none_ | No |

If you have used a common relational database, the parameters `index` and `type` can be compared to the `database` and `table` concepts.

### TLS / SSL

The Elasticsearch output plugin supports TLS/SSL. For more details about the properties available and general configuration, see [TLS/SSL](../../administration/transport-security.md).

### AWS Sigv4 Authentication and Upstream Servers

> The `http_proxy`, `no_proxy`, and `TLS` parameters used for AWS Sigv4 Authentication
> (for connection of plugin to AWS to generate authentication signature) are never
> picked from the `NODE` section of the
> [Upstream](../../administration/configuring-fluent-bit/classic-mode/upstream-servers.md)
> configuration. However, `TLS` parameters for connection of the plugin to
> Elasticsearch **can** be overridden in the `NODE` section of Upstream, even if AWS
> authentication is used.

Reviewer 3: ```suggestion
The `http_proxy`, `no_proxy`, and `TLS` parameters used for AWS Sigv4 Authentication (for connection of plugin to AWS to generate authentication signature) are never picked from the `NODE` section of the [Upstream](../../administration/configuring-fluent-bit/classic-mode/upstream-servers.md) configuration. However, `TLS` parameters for connection of the plugin to Elasticsearch can be overridden in the `NODE` section of Upstream, even if AWS authentication is used.
```

Unfortunately, GitBook messes up the rendering of properly linewrapped text :(

Author: Applied these changes

This plugin has the following configuration parameters:

> The **Overridable** column indicates if a key can be overridden in the NODE section of an

Reviewer 3: ```suggestion
The **Overrides allowed** column indicates whether a key can be overridden in the `NODE` section of an
```

Author: Changed as suggested in ea6387481e9dc2bb0f7575ec42eb7575cd5ea393.

[Upstream](../../administration/configuring-fluent-bit/classic-mode/upstream-servers.md)
configuration.

| Key | Description | Default | Overridable |
| :--- | :--- | :--- | :--- |

> | `Host` | IP address or hostname of the target Elasticsearch instance. | `127.0.0.1` | Yes. Default value isn't applicable for NODE section of Upstream configuration, which requires `Host` to be specified. |

Reviewer 3: ```suggestion
| `Host` | IP address or hostname of the target Elasticsearch instance. | `127.0.0.1` | Yes. Default value isn't applicable for `NODE` section of Upstream configuration, which requires `Host` to be specified. |
```

![example configuration visualization from calyptia](../../.gitbook/assets/image%20%282%29.png)

### Configuration File with Upstream

> In your main configuration file append the following _Input_ & _Output_ sections.

Reviewer: ```suggestion
In your main configuration file append the following `Input` and `Output` sections:
```

Author: I would prefer all cases like this to be fixed at once in a pull request dedicated to re-formatting.

```text
[INPUT]
    Name     cpu
    Tag      cpu

[OUTPUT]
    Name     es
    Match    *
    Upstream ./upstream.conf
    Index    my_index
    Type     my_type
```

> Your [Upstream Servers](../../administration/configuring-fluent-bit/classic-mode/upstream-servers.md) configuration file can look like:

Reviewer: ```suggestion
Your [Upstream Servers](../../administration/configuring-fluent-bit/classic-mode/upstream-servers.md)
configuration file can similar to the following:
```

Author: Suggestion was applied as part of f28a9c6a402fcf5dc87b1aed0902e1dff09c6749.

![example configuration visualization from Calyptia](../../.gitbook/assets/image%20%282%29.png)

### Configuration File with Upstream

In your main configuration file append the following `Input` and `Output` sections:

> ```text

Reviewer 2: We need to provide yaml config as well now by default.

Author: Added in d3181511ea54b1b2f5a823c026530915cf857887

{% endtab %}
{% endtabs %}

### Configuration File with Upstream

> #### Classic mode Configuration File with Upstream

Author: Changed as suggested in ea6387481e9dc2bb0f7575ec42eb7575cd5ea393.
